specs/ledger/hs/src/Ledger/Update.hs

```diff
@@ -539,3 +542,244 @@ instance STS UPEND where
 
 instance Embed FADS UPEND where
   wrapFailed = error "No possible failures in FADS"
+
+------------------------------------------------------------------------
+-- Update interface
+------------------------------------------------------------------------
+
+-- | The update interface environment is shared amongst various rules, so
+--   we define it as an alias here.
+type UPIEnv =
+  ( Core.Epoch
+  , Core.Slot
+  , Map Core.VKeyGenesis Core.VKey
+  )
+
+-- | The update interface state is shared amongst various rules, so we define it
+-- as an alias here.
+type UPIState =
+  ( Core.Epoch
+  , (ProtVer, PParams)
+  , [(Core.Slot, (ProtVer, PParams))]
+  , Map ApName (ApVer, Core.Slot)
+  , Map UpId (ProtVer, PParams)
+  , Map UpId (ApName, ApVer)
+  , Map UpId Core.Slot
+  , Core.PairSet UpId Core.VKeyGenesis
+  , Core.PairSet ProtVer Core.VKey
+  , Map UpId Core.Slot
+  )
+
+data UPIREG
+
+instance STS UPIREG where
+  type Environment UPIREG = UPIEnv
+  type State UPIREG = UPIState
+  type Signal UPIREG = UProp
+  data PredicateFailure UPIREG
+    = UPREGFailure (PredicateFailure UPREG)
+    deriving (Eq, Show)
+
+  initialRules = []
+  transitionRules =
+    [ do
+        TRC ( (_ec, sn, dms)
+            , ( ep
+              , (pv, pps)
+              , fads
+              , avs
+              , rpus
+              , raus
+              , cps
+              , vts
+              , bvs
+              , pws)
+            , up) <- judgmentContext
+        (rpus', raus') <- trans @UPREG $ TRC ((pv, pps, avs, dms), (rpus, raus), up)
+        let pws' = pws ⨃ [(up ^. upId, sn)]
+        return ( ep
+                , (pv, pps)
+                , fads
+                , avs
+                , rpus'
+                , raus'
+                , cps
+                , vts
+                , bvs
+                , pws'
+                )
+
+    ]
+
+instance Embed UPREG UPIREG where
+  wrapFailed = UPREGFailure
+
+data UPIVOTE
+
+instance STS UPIVOTE where
+  type Environment UPIVOTE = UPIEnv
+  type State UPIVOTE = UPIState
+  type Signal UPIVOTE = Vote
+  data PredicateFailure UPIVOTE
+    = UPVOTEFailure (PredicateFailure UPVOTE)
+    deriving (Eq, Show)
+
+  initialRules = []
+  transitionRules =
+    [ do
+        TRC ( (_ec, sn, dms)
+            , ( ep
+              , (pv, pps)
+              , fads
+              , avs
+              , rpus
+              , raus
+              , cps
+              , vts
+              , bvs
+              , pws)
+            , v) <- judgmentContext
+        (cps', vts') <- trans @UPVOTE $ TRC ((sn, pps, dom pws, dms), (cps, vts), v)
+        let avsnew = Map.fromList [(an, (av, sn)) | pid <- Map.keys cps'
```

Comment:
Can we use `dom` here instead of `Map.keys`?

Reply:
We can do, but then we'd have to convert from a set to a list. Is it worth it?

Reply:
Nah, then probably not :)